sandbox/host/src/sandbox/wasmi_backend.rs

```diff
@@ -18,393 +18,534 @@
 
 //! Wasmi specific impls for sandbox
 
-use std::fmt;
+use std::{
+    rc::{Rc, Weak},
+    slice,
+};
 
 use codec::{Decode, Encode};
-use gear_sandbox_env::HostError;
-use sandbox_wasmi::{
-    memory_units::Pages, ImportResolver, MemoryInstance, Module, ModuleInstance, RuntimeArgs,
-    RuntimeValue, Trap, TrapCode,
-};
-use sp_wasm_interface_common::{util, Pointer, ReturnValue, Value, WordSize};
+use gear_sandbox_env::{HostError, Instantiate, WasmReturnValue, GLOBAL_NAME_GAS};
+use region::{Allocation, Protection};
+use wasmi::{AsContext, AsContextMut, Engine, ExternType, Linker, MemoryType, Module, Val};
+
+use sp_wasm_interface_common::{Pointer, ReturnValue, Value, WordSize};
 
 use crate::{
     error::{self, Error},
     sandbox::{
-        BackendInstanceBundle, GuestEnvironment, GuestExternals, GuestFuncIndex, Imports,
-        InstantiationError, Memory, SandboxInstance, SupervisorContext,
+        BackendInstanceBundle, GuestEnvironment, InstantiationError, Memory, SandboxInstance,
+        SupervisorContext,
     },
+    store_refcell,
     util::MemoryTransfer,
 };
 
+use super::SupervisorFuncIndex;
+
+type Store = wasmi::Store<Option<FuncEnv>>;
+pub type StoreRefCell = store_refcell::StoreRefCell<Store>;
+
 environmental::environmental!(SupervisorContextStore: trait SupervisorContext);
 
-#[derive(Debug)]
-struct CustomHostError(String);
+pub struct FuncEnv {
+    store: Weak<StoreRefCell>,
+    gas_global: wasmi::Global,
+}
 
-impl fmt::Display for CustomHostError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "HostError: {}", self.0)
+impl FuncEnv {
+    pub fn new(store: Weak<StoreRefCell>, gas_global: wasmi::Global) -> Self {
+        Self { store, gas_global }
     }
 }
 
-impl sandbox_wasmi::HostError for CustomHostError {}
-
 /// Construct trap error from specified message
-fn trap(msg: &'static str) -> Trap {
-    Trap::host(CustomHostError(msg.into()))
+fn host_trap(msg: impl Into<error::Error>) -> wasmi::Error {
+    wasmi::Error::host(msg.into())
+}
+
+fn into_wasmi_val(value: Value) -> wasmi::Val {
+    match value {
+        Value::I32(val) => wasmi::Val::I32(val),
+        Value::I64(val) => wasmi::Val::I64(val),
+        Value::F32(val) => wasmi::Val::F32(wasmi::core::F32::from_bits(val)),
+        Value::F64(val) => wasmi::Val::F64(wasmi::core::F64::from_bits(val)),
```

Comment:
```suggestion
        Value::F64(val) => wasmi::Val::F64(f64::from_bits(val)),
```

Also, please, do the same conversions in `into_value()`

Reply:
Hmm, your suggested change doesn't even compile. Please note that `f64` and `F64` are different types.